### DISCUSSION TITLE
How to handle a feature that has related jobs

### DISCUSSION THREAD

**Discussion by yanielbf** [lucidarch/lucid#30] [Q&A]
The architecture seems incredible to me. I'm going to start using it. Thanks for the input. After reading the documentation I have doubts in relation to the characteristics. I am on the example of creating an article. I want to divide the process of creating an article into 4 jobs:
1. Validate data
2. Upload image to an amazon bucket
3. Save the post
4. Return the answer

Suppose that step 1 and 2 are executed worse when reaching three an error occurs, you would have to make a rollback of step 2 and delete the uploaded file.

Question 1: Would I have to make a job run if one of these steps cannot be created, something like DeleteImageJob?
Question 2: How would you execute an Eloquent transaction in this process?

Again, congratulations on the job. Greetings

**Comment by Batisska** [comment]
With validation, everything seems to be clear. We use standard feature requests. 

We also send the response to the task.

2 and 3 I most likely put the task in the job.

Saving and loading the image should be put in an operation and use the interface ShouldQueue.

In my humble opinion.

**Comment by Mulkave** [comment]
@yanielbf thank you for your kind words. Glad you find the architecture to be helpful.

This is a very interesting case that you mentioned and there are several approaches to handling the failure of saving and wanting a rollback.

**Method 1**
As you mentioned, at the feature level you'd create a `DeleteImage` job that runs once saving fails which would be the most obvious when looking at the feature.

**Method 2**
A little more complicated than the first one, where you'd throw an exception on failure, then the exception handler runs some sort of a "clean" job based on the type of exception that was thrown.

Answer to Q1: It is favourable that all logic is contained within jobs and operations to make code consistent and easy to navigate.
Answer to Q2: Depends on your database access code structure. For example if you access models from within Jobs, then the job should have everything needed for the save transaction and should know how to rollback and handle everything related to database communication. Otherwise, if for example you have a repository access layer, you'd include that in the repository method that is called by the job.

I hope this answers your question. Please let me know if you have any questions further.
